### ISSUE TITLE
Way to get the path, including any extension, from a target, outside its recipe

### ISSUE THREAD

**Issue by Hawk777** [build2/build2#339]
It would be nice if, given a value of type “name”, which represents a file, we could get the name of the file. Ways I have tried so far, and reasons they do not work:
* `$path.string($f)` → `error: unmatched call to path.string(<untyped>)` (reasonable, `path.string` is supposed to take a path, not a target name)
* `$target.path($f)` → `target blah not found` (reasonable, the docs say `target.path` can only be called during certain phases)
* `$name.name($f).$name.extension($f)` → works most of the time, but explodes if `$f` doesn’t have an extension due to trying to concatenate `null`; also, seems overly complicated for what should be a simple task

**Comment by Klaim** [comment]
I didn't try but I believe `$leaf($f)` or `$path.leaf($f)` should work?

**Comment by Hawk777** [comment]
If `$f` has path components in it, I want them. For example, if `$f` is `file{foo/bar/baz.txt}`, I want `foo/bar/baz.txt`, not `baz.txt`.

**Comment by boris-kolpackov** [comment]
There is no way to do this outside recipes because target path is assigned by the rule that matches the target. This "abstraction" was made to deal with extensions that vary from platform to platform (e.g., `.exe`, `.so`/`.dll`/`.dylib`).

However, if you assign the extension manually, that you can reconstruct the absolute path like so:

```
t = subdir/file{hello.txt}

p = $src_base/$directory($t)/$name($t).$extension($t) # If the target is in src.
p = $out_base/$directory($t)/$name($t).$extension($t) # If the target is in out.
```

**Comment by Hawk777** [comment]
Yeah, the problem with that is that if `$t` doesn’t *have* an extension, it explodes due to trying to concatenate `null`. My current solution is this ugliness:
```
		if $builtin.null($name.extension($f))
		{
			fn = $name.name($f)
		}
		else
		{
			fn = $name.name($f).$name.extension($f)
		}
```

**Comment by boris-kolpackov** [comment]
> if $t doesn’t _have_ an extension

This is a bit more nuanced: there is a different between a target not having an extension and the extension not being specified. Compare:

```
file{hello}   # Extension is not specified (and could be derived later by the rule).
file{hello.}  # Extension is specified as "no extension".
```

In the latter case `$extension()` will return an empty string. Which I guess will require an equally ugly `if-else` (you could probably make it a bit less ugly with `(? : )`).

We could add a function to deal with the latter case. Maybe `$name.name_with_extension()`. It could also fail with proper diagnostics if the extension if `null`.
